pytest_datadir/plugin.py

```diff
@@ -25,24 +30,149 @@ def _win32_longpath(path):
+
+    def mkdatadir(self, original_datadir='data'):
+
+        # special condition if the datadir is specified as None, which
+        # automatically gets the path that matches the basename of the
+        # module we are in
+        if original_datadir is None:
+            original_datadir = Path(os.path.splitext(self.request.module.__file__)[0])
+
+
+        # get the path to the shared data dir
+        original_path = Path(self.request.fspath.dirname) / original_datadir
+
+        # make sure that the path exists and it is a directory
+        exists = True
+        if not original_path.exists():
+            # raise the flag that it doesn't exist so we can generate
+            # a directory for it instead of copying
+            exists = False
+
+        # make sure the path is a directory if it exists
+        elif not original_path.is_dir():
+            raise ValueError("datadir path is not a directory")
+
+        # generate a base temporary directory and receive the path to it
+        temp_path = self.tmp_path_factory.mktemp('')
+
+        # in order to use the shutil.copytree util the target directory
+        # must not exist so we specify a dir in the generated tempdir for it
+        temp_data_path = temp_path / DATADIR_DIRNAME
+
+        # windows-ify the paths
+        original_path = Path(_win32_longpath(original_path))
+        temp_data_path = Path(_win32_longpath(str(temp_data_path)))
+
+        # copy or create empty directory depending on whether the
+        # original one exists
+        if exists:
+
+            # copy all the files in the original data dir to the temp
+            # dir
+            shutil.copytree(original_path, temp_data_path)
+
+        else:
+            # otherwise just give them a fallback tmpdir
+            temp_data_path.mkdir()
+
+        return temp_data_path
+
+# to clean up for the different scopes we need tmp_path_factory to be
+# scoped differently; just copy the code from module and change the
+# decorator
+
+@pytest.fixture(scope="session")
+def session_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+@pytest.fixture(scope="module")
+def module_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+
+@pytest.fixture(scope="class")
+def class_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+@pytest.fixture(scope="function")
+def function_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+# scoped factories
+@pytest.fixture(scope='module')
+def module_datadir_factory(request, module_tmp_path_factory):
```

Comment:
Same reasoning here, it doesn't make sense to have multiple scopes for datadir factories, just a session-scoped `datadir_factory` would fit the bill nicely.